src/sagemaker/git_utils.py

```diff
@@ -72,44 +81,182 @@ def git_clone_repo(git_config, entry_point, source_dir=None, dependencies=None):
+    is_ssh = git_config["repo"].startswith("git@")
+    if not is_https and not is_ssh:
+        raise ValueError("Invalid Git url provided.")
+    if is_ssh:
+        _clone_command_for_github_like_ssh(git_config, repo_dir, exists)
+    elif exists["2FA_enabled"]:
+        _clone_command_for_github_like_https_2fa_enabled(git_config, repo_dir, exists)
+    else:
+        _clone_command_for_github_like_https_2fa_disabled(git_config, repo_dir, exists)
+
+
+def _clone_command_for_github_like_ssh(git_config, repo_dir, exists):
+    if exists["username"] or exists["password"] or exists["token"]:
+        warnings.warn("Unnecessary credential argument(s) provided.")
+    _run_clone_command(git_config["repo"], repo_dir)
+
+
+def _clone_command_for_github_like_https_2fa_disabled(git_config, repo_dir, exists):
+    updated_url = git_config["repo"]
+    if exists["token"]:
+        if exists["username"] or exists["password"]:
+            warnings.warn(
+                "Using token for authentication, "
+                "but unnecessary credential argument(s) provided."
+            )
+        updated_url = _insert_token_to_repo_url(url=git_config["repo"], token=git_config["token"])
+    elif exists["username"] and exists["password"]:
+        updated_url = _insert_username_and_password_to_repo_url(
+            url=git_config["repo"], username=git_config["username"], password=git_config["password"]
+        )
+    elif exists["username"] or exists["password"]:
+        warnings.warn("Unnecessary credential argument(s) provided.")
+    _run_clone_command(updated_url, repo_dir)
+
+
+def _clone_command_for_github_like_https_2fa_enabled(git_config, repo_dir, exists):
+    updated_url = git_config["repo"]
+    if exists["token"]:
+        if exists["username"] or exists["password"]:
+            warnings.warn(
+                "Using token for authentication, "
+                "but unnecessary credential argument(s) provided."
+            )
+        updated_url = _insert_token_to_repo_url(url=git_config["repo"], token=git_config["token"])
+    elif exists["username"] or exists["password"] or exists["token"]:
+        warnings.warn(
+            "Unnecessary credential argument(s) provided."
+            "Hint: since two factor authentication is enabled, you have to provide token."
+        )
+    _run_clone_command(updated_url, repo_dir)
+
+
+def _run_clone_command(repo_url, repo_dir):
+    """Run the 'git clone' command with the repo url and the directory to clone the repo into.
+
+    Args:
+        repo_url (str): Git repo url to be cloned.
+        repo_dir: (str): Local path where the repo should be cloned into.
+
+    Raises:
+        CalledProcessError: If failed to clone git repo.
+    """
+    my_env = os.environ.copy()
+    if repo_url.startswith("https://"):
+        my_env["GIT_TERMINAL_PROMPT"] = "0"
+    elif repo_url.startswith("git@"):
+        f = tempfile.NamedTemporaryFile()
+        w = open(f.name, "w")
```

Comment:
use context manager instead:
```python
with open(...) as w
```

Find a better naming for ```w``` and ```f```

Delete f after ```subprocess.check_call```